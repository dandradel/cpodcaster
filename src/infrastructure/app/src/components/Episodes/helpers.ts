export interface TableCellI {
  id: string
  label: string
}

export const millisToMinutesAndSeconds = (millis: Date) => {
  return `${millis.getMinutes()}:${(millis.getMinutes() < 10 ? '0' : '') + millis.getMinutes()}`
}

export const tableHeadCells: TableCellI[] = [
  { id: 'title', label: 'Title' },
  { id: 'date', label: 'Date' },
  { id: 'duration', label: 'Duration' },
]
