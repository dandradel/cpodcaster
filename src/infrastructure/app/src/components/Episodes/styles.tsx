import { SxProps } from '@mui/material'

export const episodesBoxStyles: SxProps = {
  display: 'flex',
  flexDirection: 'column',
  flexBasis: '100%',
  flex: 1,
}
export const episodeNumberBoxStyles: SxProps = {
  display: 'flex',
  flexDirection: 'row',
  flexWrap: 'wrap',
  boxShadow: 3,
  minHeight: 30,
  alignItems: 'center',
  justifyContent: 'left',
  p: 1,
}
export const episodesTableStyles: SxProps = {
  display: 'flex',
  flexDirection: 'row',
  flexWrap: 'wrap',
  boxShadow: 3,
  minHeight: 30,
  alignItems: 'center',
  justifyContent: 'left',
  p: 1,
  my: 3,
}
export const tableCellStyles: SxProps = {
  fontWeight: 'bold',
}
export const tableRowStyles = {
  '&:last-child td, &:last-child th': { border: 0 },
  '&:hover': {
    cursor: 'pointer',
    backgroundColor: '#f5f5f5',
  },
}
