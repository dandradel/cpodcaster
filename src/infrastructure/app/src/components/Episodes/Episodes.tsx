import React, { ReactElement } from 'react'
import {
  Box,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from '@mui/material'
import './Episodes.styles.css'
import {
  episodesBoxStyles,
  episodeNumberBoxStyles,
  episodesTableStyles,
  tableCellStyles,
  tableRowStyles,
} from './utils'
import { NavigateFunction, useNavigate, useParams } from 'react-router-dom'
import { Episode } from '../../../../../domain/models/Episode'
import { millisToMinutesAndSeconds, TableCellI, tableHeadCells } from './helpers'

interface EpisodesProps {
  episodes: Episode[] | null
}

function Episodes({ episodes }: EpisodesProps): ReactElement {
  const { podcastId } = useParams<{ podcastId: string }>()
  const navigate: NavigateFunction = useNavigate()

  const handleEpisodeClick = (episode: Episode): void => {
    const { id } = episode
    navigate(`/podcast/${podcastId}/episode/${id}`, { state: { episode } })
  }

  const getEpisodes = (): ReactElement[] | undefined => {
    return episodes?.map((episode: Episode) => {
      const { id, title, duration, date } = episode
      const episodeDate = new Date(date).toLocaleDateString()
      const episodeTime = new Date(duration)

      return (
        <TableRow
          key={id}
          sx={tableRowStyles}
          onClick={() => handleEpisodeClick(episode)}
          data-testid='episodes-tableRow-testid'
        >
          <TableCell
            key={id}
            align='left'
            sx={{
              color: '#2E8BC0',
            }}
          >
            {title}
          </TableCell>
          <TableCell align='right'>{episodeDate}</TableCell>
          <TableCell align='right'>{millisToMinutesAndSeconds(episodeTime)}</TableCell>
        </TableRow>
      )
    })
  }

  const getTableHeadCells = (): ReactElement[] => {
    return tableHeadCells.map((tableHeadCell: TableCellI) => {
      const { id, label }: TableCellI = tableHeadCell
      return (
        <TableCell key={id} align={id === 'title' ? 'left' : 'right'} sx={tableCellStyles}>
          {label}
        </TableCell>
      )
    })
  }

  return (
    <Box sx={episodesBoxStyles} data-testid='episodes-wrapper-testid'>
      <Box sx={episodeNumberBoxStyles}>
        <Typography variant='h6' fontWeight='bold' data-testid='episodes-number-testid'>
          Episodes: {episodes?.length}
        </Typography>
      </Box>
      <Box sx={episodesTableStyles}>
        <TableContainer>
          <Table data-testid='table-episodes-testid'>
            <TableHead>
              <TableRow>{getTableHeadCells()}</TableRow>
            </TableHead>
            <TableBody>{getEpisodes()}</TableBody>
          </Table>
        </TableContainer>
      </Box>
    </Box>
  )
}

export default Episodes
