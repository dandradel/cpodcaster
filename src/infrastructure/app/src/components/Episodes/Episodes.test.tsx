import React from 'react'
import '@testing-library/jest-dom'
import { cleanup, renderWithProviders, screen } from '../../shared/test/test-utils'
import Episodes from './Episodes'
import episodesMock from '../../shared/__mocks__/episodes.json'

describe('Episodes', () => {
  afterEach(cleanup)

  test('renders Episodes', () => {
    const container = renderWithProviders(<Episodes episodes={episodesMock} />)
    expect(container).toMatchSnapshot()
  })

  test('renders Episodes with episodes number box', () => {
    renderWithProviders(<Episodes episodes={episodesMock} />)
    const episodesNumberBox = screen.getByTestId('episodes-number-testid')
    expect(episodesNumberBox).toBeInTheDocument()
  })

  test('renders Episodes with table of episodes', () => {
    renderWithProviders(<Episodes episodes={episodesMock} />)
    const episodesTable = screen.getByTestId('table-episodes-testid')
    expect(episodesTable).toBeInTheDocument()
  })
})
