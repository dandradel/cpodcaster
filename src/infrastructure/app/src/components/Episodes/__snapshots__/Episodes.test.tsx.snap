// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Episodes renders Episodes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1ixmh22"
        data-testid="episodes-wrapper-testid"
      >
        <div
          class="MuiBox-root css-4zyzsf"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-tjvwag-MuiTypography-root"
            data-testid="episodes-number-testid"
          >
            Episodes: 
            1
          </h6>
        </div>
        <div
          class="MuiBox-root css-13z1pn"
        >
          <div
            class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
          >
            <table
              class="MuiTable-root css-rqglhn-MuiTable-root"
              data-testid="table-episodes-testid"
            >
              <thead
                class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-11sptls-MuiTableCell-root"
                    scope="col"
                  >
                    Title
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1o98s2t-MuiTableCell-root"
                    scope="col"
                  >
                    Date
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1o98s2t-MuiTableCell-root"
                    scope="col"
                  >
                    Duration
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root css-1kapz33-MuiTableRow-root"
                  data-testid="episodes-tableRow-testid"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeMedium css-siow7j-MuiTableCell-root"
                  >
                    Episode 1
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
                  >
                    1/1/2020
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
                  >
                    25:25
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1ixmh22"
      data-testid="episodes-wrapper-testid"
    >
      <div
        class="MuiBox-root css-4zyzsf"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-tjvwag-MuiTypography-root"
          data-testid="episodes-number-testid"
        >
          Episodes: 
          1
        </h6>
      </div>
      <div
        class="MuiBox-root css-13z1pn"
      >
        <div
          class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
        >
          <table
            class="MuiTable-root css-rqglhn-MuiTable-root"
            data-testid="table-episodes-testid"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-11sptls-MuiTableCell-root"
                  scope="col"
                >
                  Title
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1o98s2t-MuiTableCell-root"
                  scope="col"
                >
                  Date
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1o98s2t-MuiTableCell-root"
                  scope="col"
                >
                  Duration
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            >
              <tr
                class="MuiTableRow-root css-1kapz33-MuiTableRow-root"
                data-testid="episodes-tableRow-testid"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeMedium css-siow7j-MuiTableCell-root"
                >
                  Episode 1
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
                >
                  1/1/2020
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
                >
                  25:25
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
